nubia/internal/completion.py

```diff
@@ -236,19 +236,31 @@ def _prepare_args_completions(
                 choices = arg.choices(
                     self.cmd.metadata.command.name, subcommand, last_token,
                     self.doc.text)
-                if not isinstance(choices , List):
+                if not isinstance(choices, List):
                     raise ValueError('autocomplete function MUST provide list of strings'
                                      f', got {choices}')
             else:
                 choices = arg.choices
 
-            return [
-                Completion(
-                    text=str(choice),
-                    start_position=-len(parsed_token.last_value),
-                )
-                for choice in choices
-            ]
+            if len(choices) > 10:
+                # Avoid creating slowness or problems when there are too many
+                # choices. Show just the first 10, but the rest are still
+                # available.
+                return [
+                    Completion(
+                        text=str(choice),
+                        start_position=-len(parsed_token.last_value),
+                    )
+                    for choice in choices[:11]
+                ]
+            else:
+                return [
+                    Completion(
+                        text=str(choice),
+                        start_position=-len(parsed_token.last_value),
+                    )
+                    for choice in choices
+                ]
```

Comment:
Since the return is almost identical I wouldn't repeat the code. We might exploit the fact that slicing doesn't complain if we give an out-of-bound index and do something like:
```suggestion
            # Avoid creating slowness or problems when there are too many
            # choices. Show just the first 10, but the rest are still
            # available.
            selected_choices = choices[:10]
            return [
                Completion(
                    text=str(choice),
                    start_position=-len(parsed_token.last_value),
                )
                for choice in selected_choices
            ]
```